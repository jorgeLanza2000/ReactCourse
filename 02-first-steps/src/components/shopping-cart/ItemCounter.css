.item-row {
  display: flex;
  align-items: center;
  gap: 10px;
}

.text-width {
  width: 150px;
}
