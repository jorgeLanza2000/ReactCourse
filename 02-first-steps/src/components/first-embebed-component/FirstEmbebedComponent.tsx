import { ItemCounter } from '../shopping-cart/ItemCounter';

interface ItemInCart {
  productName: string;
  quantity: number;
}

const itemsInCart: ItemInCart[] = [
  {
    productName: 'Pony station',
    quantity: 1,
  },
  {
    productName: 'Pro controller',
    quantity: 55,
  },
  {
    productName: 'Steam deck',
    quantity: 3,
  },
];

//! en lugar de @For pues es .map
//? el key no debe ser el index, mala praxis
//? obviamente key debe ser unico
export const FirstEmbebedComponent = () => {
  return (
    <section
      style={{
        display: 'flex',
        flexDirection: 'column',
        gap: 20,
      }}
    >
      <h1>Carrito de compras</h1>
      <ItemCounter name='Nintendo Switch' quantity={2} />
      <ItemCounter name='Play Station 5' quantity={1} />
      <ItemCounter name='Xbox One X' quantity={3} />
      <span>**************************************</span>
      {itemsInCart.map(({ productName, quantity }) => (
        <ItemCounter key={productName} name={productName} quantity={quantity} />
      ))}
    </section>
  );
};
