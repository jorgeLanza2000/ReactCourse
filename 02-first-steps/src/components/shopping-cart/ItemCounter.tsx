import { useState } from 'react';

//! como archivo normal de css
import './ItemCounter.css';
//! como modulo de css
import styles from './ItemCounterModule.module.css';

interface Props {
  name: string;
  quantity?: number;
}

export const ItemCounter = ({ name, quantity = 1 }: Props) => {
  //! los Hooks es lo primero de cada componente
  const [count, setCount] = useState(quantity); //* valor iniical

  const handleAdd = () => {
    setCount(count + 1);
  };
  const handleSubtract = () => {
    if (count <= 1) return;
    setCount(count - 1);
  };

  return (
    <section className='item-row'>
      <span
        style={{
          color: count === 1 ? 'red' : 'black',
        }}
        className='text-width'
      >
        {name}
      </span>
      <button onClick={handleAdd}>+1</button>
      <span
        style={{
          width: 15,
          textAlign: 'center',
        }}
        className={count === 1 ? styles.red : ''}
      >
        {count}
      </span>
      <button onClick={handleSubtract}>-1</button>
    </section>
  );
};
