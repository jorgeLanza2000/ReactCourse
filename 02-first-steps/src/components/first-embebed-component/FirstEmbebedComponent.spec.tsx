import { render, screen } from '@testing-library/react';
import { afterEach, describe, expect, test, vi } from 'vitest';
import { FirstEmbebedComponent } from './FirstEmbebedComponent';
import { ItemCounter } from '../shopping-cart/ItemCounter';

//! simulacion de funcion de retorno
const mockItemCounter = vi.fn((props: unknown) => {
  return <div data-testid='ItemCounter' />;
});

vi.mock('../shopping-cart/ItemCounter', () => ({
  ItemCounter: (props: unknown) => mockItemCounter(props),
}));

//? otra forma de hacerlo
// vi.mock('../shopping-cart/ItemCounter', () => ({
//   ItemCounter: (props: unknown) => (
//     <div
//       data-testid='ItemCounter'
//       name={props.name}
//       quantity={props.quantity}
//     />
//   ),
// }));

describe('FirstEmbebedComponent', () => {
  //! ciclos de vida a lo angular, esto es after cada test
  afterEach(() => {
    vi.clearAllMocks();
  });

  test('should match snapshot', () => {
    const { container } = render(<FirstEmbebedComponent />);
    expect(container).toMatchSnapshot(); //! es muy rigido ya que evalua la estructura html
  });

  test('should render the correct number of ItemCounter components', () => {
    render(<FirstEmbebedComponent />);
    const itemCounters = screen.getAllByTestId('ItemCounter');
    expect(itemCounters.length).toBe(6);
  });

  test('should render ItemCounter with correct props', () => {
    render(<FirstEmbebedComponent />);
    expect(mockItemCounter).toHaveBeenCalledTimes(6);
    expect(mockItemCounter).toHaveBeenCalledWith({
      name: 'Nintendo Switch',
      quantity: 2,
    });
    expect(mockItemCounter).toHaveBeenCalledWith({
      name: 'Play Station 5',
      quantity: 1,
    });
    expect(mockItemCounter).toHaveBeenCalledWith({
      name: 'Xbox One X',
      quantity: 3,
    });
    expect(mockItemCounter).toHaveBeenCalledWith({
      name: 'Pony station',
      quantity: 1,
    });
    expect(mockItemCounter).toHaveBeenCalledWith({
      name: 'Pro controller',
      quantity: 55,
    });
    expect(mockItemCounter).toHaveBeenCalledWith({
      name: 'Steam deck',
      quantity: 3,
    });
  });
});
