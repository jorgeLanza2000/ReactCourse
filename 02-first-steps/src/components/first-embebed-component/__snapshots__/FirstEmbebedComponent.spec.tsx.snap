// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstEmbebedComponent > should match snapshot 1`] = `
<div>
  <section
    style="display: flex; flex-direction: column; gap: 20px;"
  >
    <h1>
      Carrito de compras
    </h1>
    <div
      data-testid="ItemCounter"
    />
    <div
      data-testid="ItemCounter"
    />
    <div
      data-testid="ItemCounter"
    />
    <span>
      **************************************
    </span>
    <div
      data-testid="ItemCounter"
    />
    <div
      data-testid="ItemCounter"
    />
    <div
      data-testid="ItemCounter"
    />
  </section>
</div>
`;
